backend/src/user/user.route.ts

```diff
@@ -1,21 +1,70 @@
 import express from 'express'
+import { addCode, deleteCode, getAllCodes, updateCode } from './user.service'
 import { celebrate, Joi, Segments } from 'celebrate'
-import { getAllCodes } from './user.service'
 
 const router = express()
 
-router.get(
-	'/:uid/codes',
+export interface TypedRequestBody<T> extends Express.Request {
+	body: T
+}
+
+interface CodeData {
+	codename: string
+	code: string
+	input?: string
+}
+
+router.get('/codes', async (_req, res) => {
+	const { uid } = res.locals
+	const response = await getAllCodes(uid)
+	res.status(200).json(response)
+})
+
+router.post(
+	'/codes',
 	celebrate({
-		[Segments.PARAMS]: Joi.object({
-			uid: Joi.string().required(),
+		[Segments.BODY]: Joi.object().keys({
+			codename: Joi.string().required(),
+			code: Joi.string().required(),
+			input: Joi.string(),
 		}),
 	}),
-	async (req, res) => {
-		const { uid } = req.params
-		const response = await getAllCodes(uid)
+	async (req: TypedRequestBody<CodeData>, res) => {
+		const { uid } = res.locals
+		const { codename, code, input } = req.body
+		const response = await addCode(uid, codename, code, input)
+		res.status(201).json(response)
+	}
+)
+router.put(
+	'/codes',
+	celebrate({
+		[Segments.BODY]: Joi.object().keys({
+			codename: Joi.string().required(),
+			code: Joi.string().required(),
+			input: Joi.string(),
+		}),
+	}),
+	async (req: TypedRequestBody<CodeData>, res) => {
+		const { uid } = res.locals
+		const { codename, code, input } = req.body
+		const response = await updateCode(uid, codename, code, input)
 		res.status(200).json(response)
 	}
 )
+router.delete(
+	'/codes/:codename',
+	celebrate({
+		[Segments.PARAMS]: Joi.object().keys({
+			codename: Joi.string().required(),
+		}),
+	}),
+	async (req, res) => {
+		const { uid } = res.locals
+		const codename = req.params.codename as string
+		await deleteCode(uid, codename)
+		res.status(200).json({ message: 'Successful Deletion' })
+	}
+)
```

Comment:
Why is this route the only one with the format `/codes/:codename`? Typically in REST apis the `PUT` and `POST` routes follow this format as well (`/resourceName/resourceId`).

Reply:
Updated